walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -18,99 +18,154 @@
+        Collection<TimeRange> gapsWithoutOptional = findGaps(eventsWithoutOptionalAttendees, meetingTime, numAttendees);
+        return gapsWithoutOptional;
+    } else {
+        // If there is at least one valid gap that considers everyone, that is better than
+        // having more valid gaps that exclude optional attendees, so we default to returning gaps
+        // that consider everyone.
+        return gapsConsideringEveryone;
+    }
+  }
+
+  /** Discard an event if none of the attendees who want to book a meeting are going to it. */
+  private List<Event> removePeopleNotAttending(List<Event> allEvents, Collection<String> allAttendees, Collection<String> allOptionalAttendees) {
+    Iterator<Event> allEventsIter = allEvents.iterator();
+
+    while (allEventsIter.hasNext()) {
+        Event e = allEventsIter.next();
+        if (eventAttendeeNotMeeting(e, allAttendees, allOptionalAttendees)) {
+            allEventsIter.remove();
+        }
+    }
+    return allEvents;
+  }
+
+  /** Return `true` if the person attending the event isn't one of the people who wants to book a meeting. */
+  private boolean eventAttendeeNotMeeting(Event e, Collection<String> allAttendees, Collection<String> allOptionalAttendees) {
+    Set<String> attendees = e.getAttendees();
     
-    // Track latest end time of any event.
-    // Since eventTimes was sorted by start, the event with the latest end time
-    // might come before the last event in the List (e.g. if we had [5, 20]
-    // followed by [10, 15], we want to use [20, END_OF_DAY] for the final gap,
-    // not [15, END_OF_DAY]).
+    for (String s : attendees) {
+        if ( !( allAttendees.contains(s) || allOptionalAttendees.contains(s)) ) {
+            return true;
+        }
+    }
+    return false;
+  }
+
+  /** Generate list of events that excludes optional attendees. */
+  private List<Event> getEventsWithoutOptionalAttendees(List<Event> allEvents, Collection<String> allOptionalAttendees) {
+    List<Event> eventsWithoutOptionalAttendees = new ArrayList(allEvents);
+    Iterator<Event> optionalEventsIter = eventsWithoutOptionalAttendees.iterator();
+    
+    while(optionalEventsIter.hasNext()) {
+        Event e = optionalEventsIter.next();
+        for (String s : e.getAttendees()) {
+            if (allOptionalAttendees.contains(s)) {
+                optionalEventsIter.remove();
+                break;
+            }
+        }
+    }
+    return eventsWithoutOptionalAttendees;
+  }
+
+  /** Find all gaps in a given collection of event times. */
+  private Collection<TimeRange> findGaps(List<Event> events, long meetingTime, int numAttendees) {
+    // Track latest end time of any event in case the event with
+    // the latest end time comes before the last event in the event List.
+    //
+    // Example: The last free gap for events ([5, 20] , [10, 15])
+    // should be [20, END_OF_DAY], not [15, END_OF_DAY].
     int endOfLatestEvent = 0;
```

Comment:
Initialized to `TimeRange.START_OF_DAY`?

Reply:
Good point, done.